README.md
<p align="center">
  <a href="https://github.com/vendi-ai/declarai">
    <img src="https://img.shields.io/pypi/pyversions/declarai.svg" alt="versions">
  </a>
  <a href="https://github.com/vendi-ai/declarai">
    <img src="https://img.shields.io/github/license/vendi-ai/declarai.svg" alt="license">
  </a>
  <a href="https://github.com/vendi-ai/declarai/actions/workflows/test.yaml">
    <img src="https://github.com/vendi-ai/declarai/actions/workflows/test.yaml/badge.svg" alt="Tests">
  </a>
  <a href="https://pypi.org/project/declarai/">
    <img src="https://img.shields.io/pypi/v/declarai?color=%2334D058&label=pypi%20package" alt="Pypi version">
  </a>
    <a href="https://pepy.tech/project/declarai">
    <img src="https://static.pepy.tech/badge/declarai/month" alt="Pypi downloads">
  </a>
  <a href="https://discord.gg/GrszSXNTDm">
    <img src="https://dcbadge.vercel.app/api/server/GrszSXNTDm?compact=true&style=flat" alt="Discord invite">
  </a>
  <a target="_blank" href="https://colab.research.google.com/github/vendi-ai/declarai/blob/main/examples/declarai_intro.ipynb">
    <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
  </a>
</p>

<p align="center">
  <img src="assets/Logo-declarai.png" alt="Logo - declarai.png">
</p>

---

**Documentation 📖**: <a href="https://vendi-ai.github.io/declarai" target="_blank">https://declarai.com </a>

**Source Code 💻** : <a href="https://github.com/vendi-ai/declarai" target="_blank">https://github.com/vendi-ai/declarai </a>

---

## What is Declarai 🤔?

**Declarai** turns your Python code into LLM tasks, allowing you to easily integrate LLM into your existing codebase.
It operates on a simple principle: just define a Python function/class. 
By annotating this function with docstrings and type hints, you provide a clear instruction set for the AI model without any additional effort.

Once you've declared your function, Declarai intelligently compiles the function's docstrings and type hints into a prompt for the AI model, ensuring the model understands exactly what's required.

After executing the task, Declarai retrieves the AI's response and parses it, translating it back into the declared return type of your Python function. This eliminates any manual parsing or post-processing on your part.

**Declarai** Keeps It Native: At its core, Declarai is about embracing native Python practices. You don't need to learn a new syntax or adapt to a different coding paradigm. Just write Python functions as you always have, and let Declarai handle the AI integration seamlessly.

---
## Main Components 🧩

### Tasks 💡

AI tasks are used for any business logic or transformation.
```python
import declarai

gpt_35 = declarai.openai(model="gpt-3.5-turbo")

@gpt_35.task 
def rank_by_severity(message: str) -> int:
    """
    Rank the severity of the provided message by it's urgency.
    Urgency is ranked on a scale of 1-5, with 5 being the most urgent.
    :param message: The message to rank
    :return: The urgency of the message
    """


rank_by_severity(message="The server is down!")

>>> 5

rank_by_severity(message="How was your weekend?"))

>>> 1
```
### Chat 🗣

AI Chats are used for an iterative conversation with the AI model, where the AI model can remember previous messages and context.

```python
import declarai

gpt_35 = declarai.openai(model="gpt-3.5-turbo")

@gpt_35.experimental.chat
class SQLBot:
    """
    You are a sql assistant. You help with SQL related questions 
    """


sql_bot = SQLBot()
sql_bot.send("When should I use a LEFT JOIN?")

> "You should use a LEFT JOIN when you want to return all rows from ....
```

 
### Features:

- [x] 🖍 **Intelligent Prompts**: Automatically generate prompts using type hints and docstrings.
- [x] 🚄 **Conversational AI**: Chat interface equipped with memory and context management.
- [x] ⚡ **Real-time streaming**: Stream LLM responses that take longer to complete.
- [x] 🔥 **Pydantic Model Parsing**: Seamlessly parse llm responses into[ Pydantic models](https://github.com/vendi-ai/declarai#pydantic-models).
- [x] 🐍 **Pythonic**: Native understanding and parsing of llm responses into [Python primitives](https://github.com/vendi-ai/declarai#tasks-with-python-native-output-parsing).
- [x] 💾 **Multiple AI Backends**: Integrated with OpenAI & Azure AI llm providers.
- [x] 🛠 **Middleware**: Adapt and extend tasks behavior with a modular middleware system.
- [ ] 🤗 **Coming Soon**: Integration with HuggingFace hub

## Quickstart 🚀

### Installation
```bash
pip install declarai
```

### Setup
```bash
export OPENAI_API_KEY=<your openai token>
```
or pass the token when initializing the declarai object
```python
import declarai

gpt_35 = declarai.openai(model="gpt-3.5-turbo", openai_token="<your-openai-key>")
```

## 💡 Basic Usage
Craft AI-powered functionalities with ease using the `@task` decorator. Just add some type hints and a bit of documentation, and watch Declarai do its magic!
```python
import declarai

gpt_35 = declarai.openai(model="gpt-3.5-turbo")

@gpt_35.task
def generate_poem(title: str) -> str:
    """
    Write a 4 line poem on the provided title
    """


res = generate_poem(
    title="Declarai, the declarative AI framework for LLMs"
)
print(res)
# Declarai, the AI framework,
# Empowers LLMs with declarative power,
# Efficiently transforming data and knowledge,
# Unlocking insights in every hour.
```
Not the best poem out there, but hey! You've written your first declarative AI code!

Declarai aims to promote clean and readable code by enforcing the use of doc-strings and typing.
The resulting code is readable and easily maintainable.


### Tasks with python native output parsing

Python primitives
```python
import declarai

gpt_35 = declarai.openai(model="gpt-3.5-turbo")

@gpt_35.task
def rank_by_severity(message: str) -> int:
    """
    Rank the severity of the provided message by it's urgency.
    Urgency is ranked on a scale of 1-5, with 5 being the most urgent.
    :param message: The message to rank
    :return: The urgency of the message
    """


rank_by_severity(message="The server is down!")

>>> 5
rank_by_severity(message="How was your weekend?"))

>>> 1
```
Python Lists/Dicts etc..

```python
from typing import List
import declarai

gpt_35 = declarai.openai(model="gpt-3.5-turbo")

@gpt_35.task
def multi_value_extraction(text: str) -> List[str]:
    """
    Extract the phone numbers from the provided text
    :param text: content to extract phone number from
    :return: The phone numbers that where identified in the input text
    """


multi_value_extraction(
    text="Hey jenny,\nyou can call me at 124-3435-132.\n"
         "you can also reach me at +43-938-243-223"
)
>>> ['124-3435-132', '+43-938-243-223']
```
Python complex objects
```python
from datetime import datetime
import declarai

gpt_35 = declarai.openai(model="gpt-3.5-turbo")

@gpt_35.task
def datetime_parser(raw_date: str) -> datetime:
    """
    Parse the input into a valid datetime string of the format YYYY-mm-ddThh:mm:ss
    :param raw_date: The provided raw date
    :return: The parsed datetime output
    """


datetime_parser(raw_date="Janury 1st 2020"))

>>> 2020-01-01 00:00:00
```

### Pydantic models
```python
from pydantic import BaseModel
from typing import List, Dict
import declarai


class Animal(BaseModel):
    name: str
    family: str
    leg_count: int

gpt_35 = declarai.openai(model="gpt-3.5-turbo")

@gpt_35.task
def suggest_animals(location: str) -> Dict[int, List[Animal]]:
    """
    Create a list of numbers from 0 to 5
    for each number, suggest a list of animals with that number of legs
    :param location: The location where the animals can be found
    :return: A list of animal leg count and for each count, the corresponding animals
    """


suggest_animals(location="jungle")

>>> {
       0: [
           Animal(name='snake', family='reptile', leg_count=0)
       ], 
       2: [
           Animal(name='monkey', family='mammal', leg_count=2), 
           Animal(name='parrot', family='bird', leg_count=2)
       ], 
       4: [
          Animal(name='tiger', family='mammal', leg_count=4), 
          Animal(name='elephant', family='mammal', leg_count=4)
       ]
 }
```
### Jinja templates 🥷
```python
import declarai

gpt_35 = declarai.openai(model="gpt-3.5-turbo")

@gpt_35.task
def sentiment_classification(string: str, examples: List[str, int]) -> int:
    """
    Classify the sentiment of the provided string, based on the provided examples.
    The sentiment is ranked on a scale of 1-5, with 5 being the most positive.
    {% for example in examples %}
    {{ example[0] }} // {{ example[1] }}
    {% endfor %}
    {{ string }} //
    """
    
sentiment_classification(string="I love this product but there are some annoying bugs",
                         examples=[["I love this product", 5], ["I hate this product", 1]])

>>> 4
```

### Simple Chat interface

```python
import declarai

gpt_35 = declarai.openai(model="gpt-3.5-turbo")

@gpt_35.experimental.chat
class CalculatorBot:
    """
    You a calculator bot,
    given a request, you will return the result of the calculation
    """

    def send(self, message: str) -> int: ...


calc_bot = CalculatorBot()
calc_bot.send(message="1 + 1")

>>> 2
```


📚 For a thorough introduction, features, and best practices, explore our [official documentation](https://vendi-ai.github.io/declarai/) and [beginner's guide](https://vendi-ai.github.io/declarai/beginners-guide/).

## Contributing 💼
Join our mission to make declarative AI even better together! Check out our [contributing guide](https://vendi-ai.github.io/declarai/contribute/) to get started.



docs/changelog.md

---
### v0.1.10

[View full release on GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.10) and [PyPi](https://pypi.org/project/declarai/0.1.10/)

**Streaming, Docstring Templates, and System Message Fixes**

New features around streaming for chat tasks and the use of Jinja templates in docstrings, plus a crucial fix for system messages in chat.

Features:

  - Streaming for chat tasks.
  - Support for Jinja templates in docstrings.

Fixes:

  - System message correctly set when passed through the chat decorator.

Changes:

  - Added streaming support with middleware for chat tasks as well. Closes #112.
  - Typos in README fixed.
  - Jinja templates in docstrings supported. Closes #118.
  - Fixed issue with passing system messages in chat decorator. Closes #121.

Contributors:

  - @matankley

---
### v0.1.9

[View full release on GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.9) and [PyPi](https://pypi.org/project/declarai/0.1.9/)

**Streaming, Middleware, and API Refactoring**

Introducing the support for streaming responses and middlewares in Declarai chats, along with API improvements.

Features:

  - Streaming responses for Declarai tasks enabled.
  - Middleware support for Declarai chats.
  
Changes:

  - Introduced middleware support to chat. Closes #105.
  - Added streaming for OpenAI tasks. Closes #107.
  - API flattened and refactored, thanks to @giladbarnea.
  
New Contributors:

  - @giladbarnea made their first contribution in #104

Contributors:

  - @giladbarnea
  - @matankley

---

### v0.1.8

[View full release on GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.8) and [PyPi](https://pypi.org/project/declarai/0.1.8/)

Changes:

  - Dual support for DECLARAI_OPEN_AI_KEY and OPENAI_API_KEY.
  - Introduced support for pydantic description field.
  - Enabled Azure OpenAI LLM. Closes #67 and #72.
  - Default Azure API version fixed.

Contributors:

  - @helmanofer
  - @matankley
  - @shobhit9957

---
### v0.1.7

[View full release on GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.7) and [PyPi](https://pypi.org/project/declarai/0.1.7/)

**Improvements and Refactoring**

A major focus on refining API Reference docs and refactoring the internal structure of the project.

Changes:

  - Introduction of gpt-3.5-turbo-16k support.
  - Internal structure refactored and API reference generation in the docs.
  - Fixes and updates in documentation.

Contributors:

  - @helmanofer
  - @matankley

---


### v0.1.6

[View full release on GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.6) and [PyPi](https://pypi.org/project/declarai/0.1.6/)

**Bug Fix Release**

Changes:

  - Fixed chat greeting duplication bug. Closes #84.

Contributors:

  - @matankley

---

### v0.1.5

[View full release on GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.5) and [PyPi](https://pypi.org/project/declarai/0.1.5/)

**Extended Database Support**

In this update, Declarai extends its support to PostgreSQL, Redis, and MongoDB databases for saving chat message history.

Changes:

  - Support for PostgreSQL, Redis, and MongoDB databases to save chat messages. Closes #77.

Contributors:

  - @matankley

---

### v0.1.4

[View full release on GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.4) and [PyPi](https://pypi.org/project/declarai/0.1.4/)

**Introduction of Chat Memory**

Declarai now supports retaining message history across chat sessions, ensuring continuity and context.

Changes:

  - Removal of operator inheritance from llm params type.
  - Added chat memory to preserve message history between sessions. Closes #62.

Contributors:

  - @matankley

---


### v0.1.3

[View full release on GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.3) and [PyPi](https://pypi.org/project/declarai/0.1.3/)

**Enhanced Control Over LLM Params**

In this version, users gain the ability to finetune the parameters of the Language Model like temperature and max tokens. 

Changes:

  - Control the llm params such as temperature, max tokens, and more. Closes #54.
  - Deployment example with FastAPI for Declarai tasks added. Closes #63.
  - Support for passing LLM parameters directly to tasks. Closes #70.
  - Comprehensive documentation added for LLMParams control.
  - Updated package version.

Contributors:

  - @matankley

---


### v0.1.2

View full release on [GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.2) and [PyPi](https://pypi.org/project/declarai/0.1.2/)

**Minor bug fixes**

Changes:

  - Updates to documentation
  - Updates to dependencies with reported vulnerabilities
  - Fix typing and improve support for IDE autocompletion
  - Fix issue with initialization failing when passed the `openai_token` at runtime.




### v0.1.1

View full release on [GitHub](https://github.com/vendi-ai/declarai/releases/tag/v0.1.1) and [PyPi](https://pypi.org/project/declarai/0.1.1/)

**Announcing the first release of Declarai!** 🥳 🥳

Declarai was born out of the awe and excitement of LLMs, along with our passion for excellent engineering and real-world applications at scale.

We hope this project will help introduce more developers into the world of LLMs and enable them to more easily and reliably integrate these amazing capabilities into their production systems.

Main features:

  - Task interface
  - Chat interface
  - Middlewares
  - Exhaustive documentation
  - OpenAI support
  - LLM prompt best practices


experiment.py
import datetime
from typing import Dict, Callable

from wandb_addons.prompts import Trace

from declarai import Declarai, Sequence
from declarai.tasks.llm_task import LLMTaskType

declarai = Declarai(provider="openai", model="gpt-3.5-turbo", version="0301")


@declarai.task
def data_manipulation(data: Dict[str, str]) -> Dict[str, str]:
    """
    create a redacted version of the input data.
    any potentially private data in the input
    should be replaced with the first 2 characters of the original content
    followed by * for the entire of the remaining characters
    for example, "guy.b@company.com" should be redacted to "gu**************"
    Don't write any python code, only return the resulting json
    :param data: The data to anonymize
    :return: The anonymized data
    """
    return declarai.magic("redacted_info", data=data)


def redact_content(task: LLMTaskType) -> Callable:
    def redact(text: str):
        sequence = Sequence(
            data_manipulation.plan(data=task.plan(text=text)),
            reduce_strategy="CoT",
        )
        return sequence()['redacted_info']

    return redact


def log_before(task: LLMTaskType) -> Callable:
    task._start_time_ms = datetime.datetime.now().timestamp() * 1000


def log_after(task: LLMTaskType) -> Callable:
    status = "success"
    status_message = None,
    # response_text = response["choices"][0]["message"]["content"]
    response_text = task._result
    # token_usage = response["usage"].to_dict()
    token_usage = 100
    end_time_ms = round(datetime.datetime.now().timestamp() * 1000)  # logged in milliseconds
    root_span = Trace(
        name="root_span",
        kind="llm",  # kind can be "llm", "chain", "agent" or "tool"
        status_code=status,
        status_message=status_message,
        metadata={"temperature": task._prompt_config.temperature,
                  "token_usage": token_usage,
                  "model_name": task._llm.model},
        start_time_ms=task._start_time_ms,
        end_time_ms=end_time_ms,
        inputs={"query": "Something"},
        outputs={"response": response_text},
    )

    # log the span to wandb
    root_span.log(name="openai_trace")


@declarai.task(
    before_hooks=[log_before],
    after_hooks=[log_after],
)
def extract_info(text: str) -> Dict[str, str]:
    """
    Extract the phone number, name and email from the provided text
    :param text: content to extract the info from
    :return: The info extracted from the text
    """
    return Declarai.magic(text=text)


res = extract_info(text="Hey jenny,"
                        "you can call me at 124-3435-132."
                        "You can also email me at georgia@coolmail.com"
                        "Have a great week!")
# print(res)
# {
#     'phone_number': '12*********',
#     'name': 'je****',
#     'email': 'ge**************'
# }

#
# import wandb
#
# # start a new wandb run to track this script
# wandb.init(
#     # set the wandb project where this run will be logged
#     project="llm-prompt-tracking",
#
#     # track hyperparameters and run metadata
#     config={
#         "learning_rate": 0.02,
#         "architecture": "CNN",
#         "dataset": "CIFAR-100",
#         "epochs": 10,
#     }
# )


    # except Exception as e:
    #     end_time_ms = round(datetime.datetime.now().timestamp() * 1000)  # logged in milliseconds
    #     status = "error"
    #     status_message = str(e)
    #     response_text = ""
    #     token_usage = {}
